### PR TITLE
fix(reports): Clear last value when state is WORKING

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

This PR fixes an issue with the alert feature, specifically the execution log reports different alert values for different states within the same execution run. Per [this](https://apache-superset.slack.com/archives/G013HAE6Y0K/p1651543102805559) Slack thread the underlying issue is the report schedule synopsis—from which the execution log is created, i.e., a point-in-time snapshot—is when the state is `WORKING` the `last_value` and `last_value_row_json` need to be cleared otherwise they incorrectly reflect the values of the previous execution run.

I’m actually not sure whether there’s any value logging the `WORKING` state to the execution log. Does it provide any relevant information to the user?

#### Before Migration

Screen Shot 2022-05-03 at 8 06 09 PM

#### After Migration

Screen Shot 2022-05-03 at 8 06 37 PM

### TESTING INSTRUCTIONS

Added unit tests and migration. The migration successfully ran in under a minute on a replica of Airbnb's production database with ~ 50,000 execution logs.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/reports/commands/execute.py [hunk 1]
```diff
@@ -94,35 +94,39 @@ def __init__(
         self._start_dttm = datetime.utcnow()
         self._execution_id = execution_id
 
-    def set_state_and_log(
+    def update_report_schedule_and_log(
```
Comment (john-bodley) [inline superset/reports/commands/execute.py:97]: I felt that `update_report_schedule_and_log` was more reflective of the function logic.

File: superset/reports/commands/execute.py [hunk 2]
```diff
@@ -94,35 +94,39 @@ def __init__(
...
+        execution log.
         """
+
+        if state == ReportState.WORKING:
+            self._report_schedule.last_value = None
+            self._report_schedule.last_value_row_json = None
+
         self._report_schedule.last_state = state
-        self._report_schedule.last_eval_dttm = dttm
+        self._report_schedule.last_eval_dttm = datetime.utcnow()
```
Comment (john-bodley) [inline superset/reports/commands/execute.py:124]: There's no need to pass in the timestamp which isn't used elsewhere.

File: superset/reports/commands/execute.py [hunk 3]
```diff
@@ -132,7 +136,7 @@ def create_log(
             end_dttm=datetime.utcnow(),
             value=self._report_schedule.last_value,
             value_row_json=self._report_schedule.last_value_row_json,
-            state=state,
+            state=self._report_schedule.state,
```
Comment (john-bodley) [inline superset/reports/commands/execute.py:139]: There's no need to pass in the state given the report schedule record has been updated with the new state.

**Comment by ktmud** [comment]
Is the report scheduler currently adding two execution logs for every execution, one in WORKING state, one in COMPLETE? Was the original intention to update the current WORKING execution to COMPLETE once the execution is done?

File: superset/migrations/versions/cbe71abde154_fix_report_schedule_and_log.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    for schedule in (
+        session.query(ReportSchedule)
+        .filter(ReportSchedule.last_state == ReportState.WORKING)
+        .all()
+    ):
+        schedule.last_value = None
+        schedule.last_value_row_json = None
```
Comment (eschutho) [inline superset/migrations/versions/cbe71abde154_fix_report_schedule_and_log.py:67]: @john-bodley do you think there's a chance that if there is a report that's in a working state, that we could potentially wind up altering the value after the report has completed? Just based off your migration timing of around a minute and most of these reports take under a minute to run, we may wind up with some reports that wind up with a completed state and no value, i.e., the commit happens here after the state has changed on the report side. Just wondering if it's safer not to change reports that are in progress.

**Review by eschutho — APPROVED** [review]
Looks great. Just one question, but not a blocker.

**Comment by eschutho** [comment]
> Is the report scheduler currently adding two execution logs for every execution, one in WORKING state, one in COMPLETE? Was the original intention to update current WORKING execution to COMPLETE (or FAILED) once an execution is done?

No, the logs are immutable, so you'll get two logs, one for each state.

**Comment by eschutho** [comment]
>I’m actually not sure whether there’s any value logging the WORKING state to the execution log. Does it provide any relevant information to the user?
@john-bodley I suppose it could give the user a start time. It's usually the same as the scheduled time, but if the report is taking a long time to run, it could indicate to the user that something is happening.

**Comment by ktmud** [comment]
> No, I don't believe this was the intention.

Then I don't understand why we have an [`end_dttm`](https://github.com/apache/superset/blob/6903c48c96fac555ca08fef3255c54522479f387/superset/models/reports.py#L204) (and "duration") for each log? If we are not updating the log record, i.e., it is a "fire and forgot" operation, shouldn't a `start_ddm` be enough? The log list table is also less readable when you have duplicate records with one having empty values and one having a spinner icon that never changes state.

**Comment by john-bodley** [comment]
I think this PR makes things less wrong so it's directionally correct. There's likely merit in rethinking what the desirable UX is from a reporting standpoint, though that can be handled in a different PR.

File: superset/migrations/versions/cbe71abde154_fix_report_schedule_and_log.py (hunk 4 above)
Comment (john-bodley) [reply superset/migrations/versions/cbe71abde154_fix_report_schedule_and_log.py:67]: @eschutho I guess with any migration there's a risk of a race condition happening whilst the migration is running. I've replicated the `session.commit()` to ensure that there's also a commit directly after the `ReportSchedule` which likely only has a handful of reports in `WORKING` state.

**Pull request merged**